engine/server/sv_client.c

```diff
@@ -2119,6 +2119,780 @@ static qboolean SV_SendBuildInfo_f( sv_client_t *cl )
+				SV_FreeEdict( ent );
+			ent = NULL;
+		}
+	}
+
+	// XashXT does not implement SV_CreateEntity, use saverestore export
+	if( !ent && svgame.physFuncs.pfnCreateEntitiesInRestoreList )
+	{
+		SAVERESTOREDATA data = { 0 };
+		ENTITYTABLE table = { 0 };
+		data.tableCount = 1;
+		data.pTable = &table;
+		table.classname = classname;
+		table.id = -1;
+		table.size = 1;
+		svgame.physFuncs.pfnCreateEntitiesInRestoreList( &data, 0, false );
+		ent = table.pent;
+	}
+
+	if( !ent )
+	{
+		SV_ClientPrintf( cl, "Invalid entity!\n" );
+		return false;
+	}
+
+	// choose default origin
+	ent->v.origin[2] = cl->edict->v.origin[2] + 25;
+	ent->v.origin[1] = cl->edict->v.origin[1] + 100 * sin( DEG2RAD( cl->edict->v.angles[1] ) );
+	ent->v.origin[0] = cl->edict->v.origin[0] + 100 * cos( DEG2RAD( cl->edict->v.angles[1] ) );
+
+	SV_LinkEdict( ent, false );
+
+	// apply keyvalues if supported
+	if( svgame.dllFuncs.pfnKeyValue )
+	{
+		for( i = 2; i < Cmd_Argc() - 1; i++ )
+		{
+			string keyname;
+			string value;
+			KeyValueData pkvd;
+
+			// allow split keyvalues to prespawn and postspawn
+			if( !Q_strcmp( Cmd_Argv( i ), "|" ) )
+				break;
+
+			Q_strncpy( keyname, Cmd_Argv( i++ ), sizeof( keyname ) - 1 );
+			Q_strncpy( value, Cmd_Argv( i ), sizeof( value ) - 1);
+			pkvd.fHandled = false;
+			pkvd.szClassName = (char*)STRING( ent->v.classname );
+			pkvd.szKeyName = keyname;
+			pkvd.szValue = value;
+			svgame.dllFuncs.pfnKeyValue( ent, &pkvd );
+
+			if( pkvd.fHandled )
+				SV_ClientPrintf( cl, "value \"%s\" set to \"%s\"!\n", pkvd.szKeyName, pkvd.szValue );
+		}
+	}
+
+	// set default targetname
+	if( !ent->v.targetname )
+	{
+		string newname, clientname;
+		int j;
+
+		for( j = 0; j < sizeof( cl->name ); j++ )
+		{
+			char c = Q_tolower( cl->name[j] );
+			if( c < 'a' || c > 'z' )
+				c = '_';
+			if( !cl->name[j] )
+			{
+				clientname[j] = 0;
+				break;
+			}
+			clientname[j] = c;
+		}
+
+		// generate name based on nick name and index
+		Q_snprintf( newname, sizeof( newname ) - 1, "%s_%i_e%i", clientname, cl->userid, NUM_FOR_EDICT( ent ));
+
+		// i know, it may break strict aliasing rules
+		// but we will not lose anything in this case.
+		Q_strnlwr( newname, newname, sizeof( newname ) - 1 );
+		ent->v.targetname = ALLOC_STRING( newname );
+		SV_EntSendVars( cl, ent );
+	}
+
+	SV_ClientPrintf( cl, "Created %i: %s, targetname %s\n", NUM_FOR_EDICT( ent ), Cmd_Argv( 1 ), STRING( ent->v.targetname ) );
+
+	if( svgame.dllFuncs.pfnSpawn )
+		svgame.dllFuncs.pfnSpawn( ent );
+
+	// now drop entity to floor.
+	pfnDropToFloor( ent );
+
+	// force think. Otherwise given weapon may crash server if player touch it before.
+	svgame.dllFuncs.pfnThink( ent );
+	pfnDropToFloor( ent );
+
+	// apply postspawn keyvales if supported
+	if( svgame.dllFuncs.pfnKeyValue )
+	{
+		for( i = i + 1; i < Cmd_Argc() - 1; i++ )
+		{
+			string keyname;
+			string value;
+			KeyValueData pkvd;
+
+			Q_strncpy( keyname, Cmd_Argv( i++ ), sizeof( keyname ) - 1 );
```

Comment:
> sizeof( keyname ) - 1

It's incorrect because Q_strncpy implemented as [strlcpy](https://man.openbsd.org/strlcpy)
So you don't need to reduce size limit.

Reply:
Ok, I fixed this